packages/astro/src/core/server-islands/vite-plugin-server-islands.ts

```diff
@@ -2,44 +2,92 @@ import type { AstroPluginMetadata } from '../../vite-plugin-astro/index.js';
 	return {
 		name: 'astro:server-islands',
 		enforce: 'post',
 		configureServer(_server) {
 			viteServer = _server;
 		},
+		resolveId(name) {
+			if(name === VIRTUAL_ISLAND_MAP_ID) {
+				return RESOLVED_VIRTUAL_ISLAND_MAP_ID;
+			}
+		},
+		load(id) {
+			if(id === RESOLVED_VIRTUAL_ISLAND_MAP_ID) {
+				return `export const serverIslandMap = ${serverIslandPlaceholder};`;
+			}
+		},
 		transform(code, id, options) {
 			if(id.endsWith('.astro')) {
 				const info = this.getModuleInfo(id);
 				if(info?.meta) {
 					const astro = info.meta.astro as AstroPluginMetadata['astro'] | undefined;
 					if(astro?.serverComponents.length) {
-						if(viteServer) {
-							for(const comp of astro.serverComponents) {
-								if(!settings.serverIslandNameMap.has(comp.resolvedPath)) {
-									let name = comp.localName;
-									let idx = 1;
-
-									while(true) {
-										// Name not taken, let's use it.
-										if(!settings.serverIslandMap.has(name)) {
-											break;
-										}
-										// Increment a number onto the name: Avatar -> Avatar1
-										name += idx++;
+						for(const comp of astro.serverComponents) {
+							if(!settings.serverIslandNameMap.has(comp.resolvedPath)) {
+								let name = comp.localName;
+								let idx = 1;
+
+								while(true) {
+									// Name not taken, let's use it.
+									if(!settings.serverIslandMap.has(name)) {
+										break;
 									}
-									settings.serverIslandNameMap.set(comp.resolvedPath, name);
-									settings.serverIslandMap.set(name, () => {
-										return viteServer?.ssrLoadModule(comp.resolvedPath) as any;
+									// Increment a number onto the name: Avatar -> Avatar1
+									name += idx++;
+								}
+
+								// Append the name map, for prod
+								settings.serverIslandNameMap.set(comp.resolvedPath, name);
+
+								settings.serverIslandMap.set(name, () => {
+									return viteServer?.ssrLoadModule(comp.resolvedPath) as any;
+								});
+
+								// Build mode
+								if(!viteServer) {
+									let referenceId = this.emitFile({
+										type: 'chunk',
+										id: comp.specifier,
+										importer: id,
+										name: comp.localName
 									});
+
+									referenceIdMap.set(comp.resolvedPath, referenceId);
 								}
 							}
 						}
 					}
 				}
 			}
+		},
+		generateBundle(options, bundles) {
+			let mapSource = 'new Map([';
+			for(let [resolvedPath, referenceId] of referenceIdMap) {
+				const fileName = this.getFileName(referenceId);
+				const islandName = settings.serverIslandNameMap.get(resolvedPath)!;
+				mapSource += `\n\t['${islandName}', () => import('./${fileName}')],`
+			}
+			mapSource += '\n]);';
+			referenceIdMap.clear();
+
+			for (const [fileName, output] of Object.entries(bundles)) {
+				if(output.type !== 'chunk') continue;
+
+				//console.log("OUTPUT", output.code);
+
+				if(output.code.includes(serverIslandPlaceholder)) {
+					output.code = output.code.replace(serverIslandPlaceholder, mapSource);
+				}
```

Comment:
Editing the code in place wouldn't update the file hash anymore, which could lead to subtle bugs. I think you can use the `augmentChunkHash` rollup hook to add the relevant reference ids as keys for Rollup to hash to fix this.

Reply:
I used `renderChunk`